File: controller/rimbot/food_forecast.py
"""Conservative per-colonist runway under observed demand and rot deadlines."""
from math import inf, isfinite


def food_forecast(supply):
    """Split shared nutrition by demand; never lend another pawn's inventory.

    Each pawn consumes its allocated stocks in earliest-expiry order. This is a
    feasible food allocation under fixed temperature/access, not a prediction of
    actual job selection. Invalid or truncated inputs cannot establish runway.
    """
    try:
        if not isinstance(supply, dict) or supply.get('readable') is not True or supply.get('truncated'):
            raise ValueError('Food supply observation is unavailable or truncated')
        consumers = supply['consumers']
        demand = {}
        for row in consumers:
            identity, rate = row['id'], number(row['nutritionPerDay'])
            if not isinstance(identity, str) or not identity or identity in demand or rate < 0:
                raise ValueError('Invalid or duplicate food consumer')
            demand[identity] = rate
        total = sum(demand.values())
        if not total:
            raise ValueError('No positive native food demand')
        stocks, seen = [], set()
        for row in supply['stocks']:
            identity, amount, holder = row['id'], number(row['nutrition']), row['holder']
            if not isinstance(identity, str) or not identity or identity in seen or amount < 0:
                raise ValueError('Invalid or duplicate food stock')
            if holder is not None and holder not in demand:
                raise ValueError('Food holder is not an observed consumer')
            seen.add(identity)
            if row['perishable'] is True:
                expiry = number(row['rotTicks']) / 60000
                if expiry < 0:
                    raise ValueError('Negative rot deadline')
            elif row['perishable'] is False:
                expiry = inf
            else:
                raise ValueError('Food perishability is unavailable')
            stocks.append((expiry, identity, amount, holder))
        stocks.sort()
        rows = []
        for identity, rate in demand.items():
            if rate == 0:
                continue
            elapsed = consumed = allocated = 0.
            for expiry, _, amount, holder in stocks:
                if holder is not None and holder != identity:
                    continue
                share = amount if holder == identity else amount * rate / total
                allocated += share
                usable = min(share, max(0., expiry - elapsed) * rate)
                consumed += usable
                elapsed += usable / rate
            rows.append({'id': identity, 'runwayDays': elapsed, 'usableNutrition': consumed,
                         'allocatedNutrition': allocated, 'nutritionPerDay': rate})
        return {'readable': True, 'runwayDays': min(row['runwayDays'] for row in rows),
                'usableNutrition': sum(row['usableNutrition'] for row in rows),
                'atRiskNutrition': sum(row['allocatedNutrition'] - row['usableNutrition'] for row in rows),
                'inventoryNutrition': sum(amount for _, _, amount, holder in stocks if holder is not None),
                'consumers': rows, 'assumptions': supply.get('assumptions', [])}
    except (KeyError, TypeError, ValueError) as error:
        return {'readable': False, 'runwayDays': None, 'reason': str(error)}


def number(value):
    if type(value) not in (int, float) or not isfinite(value):
        raise ValueError('Native food quantity is unavailable')
    return value


def acquisition_targets(facts, target_days, *, limit=8):
    """Bound new harvest orders by observed nutrition deficit and pending yield.

    Outstanding designations limit additional acquisition but never count as
    stored food or clear the food goal. One indivisible plant may overshoot.
    """
    demand = number(facts.get('nutritionPerDay'))
    forecast = facts.get('foodForecast')
    if forecast is not None:
        if forecast.get('readable') is not True:
            raise ValueError('Food forecast is unavailable')
        deficit = sum(max(0., number(row['nutritionPerDay']) * target_days - number(row['usableNutrition']))
                      for row in forecast['consumers'])
    else:
        deficit = max(0., demand * target_days - number(facts.get('foodNutrition')))
    candidates = [p for p in facts.get('acquisition', []) if p.get('food')]
    outstanding = number(facts['pendingFoodNutrition']) if 'pendingFoodNutrition' in facts else sum(
        number(p.get('nutritionYield')) for p in candidates if p.get('designated'))
    needed = max(0., deficit - outstanding)
    selected, expected = [], 0.
    for plant in candidates:
        if plant.get('designated') or len(selected) >= limit or expected >= needed:
            continue
        amount = number(plant.get('nutritionYield'))
        if amount <= 0:
            continue
        selected.append(plant)
        expected += amount
    return selected, {'targetDays': target_days, 'neededNutrition': needed,
                      'outstandingNutrition': outstanding, 'selectedNutrition': expected}

File: go/internal/policy/food_forecast.go
package policy

import (
	"errors"
	"math"
	"sort"
	"strings"
	"unicode/utf8"

	"github.com/davidarcher/RimGovernor/go/internal/domain"
)

var ErrFoodFacts = errors.New("food forecast inputs unavailable or invalid")

type FoodConsumer struct {
	ID              PawnID
	NutritionPerDay domain.Fact[float64]
	// Includes animals whose native diet accepts it. Unknown is not consent.
	HumanMeatAcceptable domain.Fact[bool]
}
type FoodStock struct {
	ID string
	// Includes meals containing humanlike meat, not just raw meat.
	IsHumanMeat bool
	RawMeat     bool
	IsHumanlike bool
	Vegetable   bool
	RawClass    domain.Fact[FoodIngredientClass]
	// Reserve is forbidden pemmican or survival meals; eligibility is after release.
	Reserve bool
	// Known empty holder means shared stock. Unknown ownership is not shared.
	Holder     domain.Fact[PawnID]
	Nutrition  domain.Fact[float64]
	Eaters     []PawnID
	Perishable domain.Fact[bool]
	RotTicks   domain.Fact[int64]
	// DefName and Count are only needed to select a specific feed resource
	// (MaintainAnimalFeed's update_feed_goal port); ForecastFood itself never
	// reads them, so unknown/zero values never affect forecast validity.
	DefName Resource
	Count   domain.Fact[int64]
	// Roofed, TemperatureC and Room describe where the stock sits at
	// observation time; ForecastFood ignores them, storage upkeep and
	// refrigeration policy read them.
	Roofed       domain.Fact[bool]
	TemperatureC domain.Fact[float64]
	Room         domain.Fact[string]
	// Corpse nutrition is its native meat yield. Forbidden corpses remain
	// observable reserves but do not extend the food runway.
	Corpse        bool
	Forbidden     domain.Fact[bool]
	MeatAmount    domain.Fact[float64]
	BodySize      domain.Fact[float64]
	TileFootprint domain.Fact[int64]
}
type FoodSupply struct {
	Complete  domain.Fact[bool]
	Consumers []FoodConsumer
	Stocks    []FoodStock
	Larder    domain.Fact[FoodLarder]
}
type ConsumerFoodForecast struct {
	ID                                                               PawnID
	RunwayDays, UsableNutrition, AllocatedNutrition, NutritionPerDay float64
}
type FoodForecast struct {
	RunwayDays                                           domain.Fact[float64]
	UsableNutrition, AtRiskNutrition, InventoryNutrition float64
	Consumers                                            []ConsumerFoodForecast
}

func foodNumber(n float64) bool { return !math.IsNaN(n) && !math.IsInf(n, 0) && n >= 0 }
func foodID(id string) bool {
	return len(id) > 0 && len(id) <= 256 && utf8.ValidString(id) && strings.TrimSpace(id) != "" && !strings.ContainsRune(id, 0)
}

// ForecastFood allocates each stock among its observed eligible consumers by
// demand, then consumes allocations in expiry order. Private inventory belongs
// only to its holder. A selected human census still competes with animal demand.
// This is feasible nutrition under fixed access/temperature, not predicted jobs.
// Nil selected means all consumers; an empty selection has unknown runway.
func ForecastFood(supply FoodSupply, selected []PawnID) (FoodForecast, error) {
	fail := func() (FoodForecast, error) { return FoodForecast{}, ErrFoodFacts }
	if complete, known := supply.Complete.Value(); !known || !complete || len(supply.Consumers) > 256 || len(supply.Stocks) > 4096 {
		return fail()
	}
	demand := map[PawnID]float64{}
	humanMeat := map[PawnID]bool{}
	ids := make([]PawnID, 0, len(supply.Consumers))
	for _, consumer := range supply.Consumers {
		rate, known := consumer.NutritionPerDay.Value()
		if _, exists := demand[consumer.ID]; exists || !foodID(string(consumer.ID)) || !known || !foodNumber(rate) {
			return fail()
		}
		demand[consumer.ID] = rate
		humanMeat[consumer.ID], _ = consumer.HumanMeatAcceptable.Value()
		ids = append(ids, consumer.ID)
	}
	sort.Slice(ids, func(i, j int) bool { return ids[i] < ids[j] })
	var total float64
	for _, id := range ids {
		total += demand[id]
	}
	if !foodNumber(total) || total == 0 {
		return fail()
	}
	wanted := map[PawnID]bool{}
	if selected == nil {
		selected = ids
	}
	for _, id := range selected {
		if _, exists := demand[id]; !exists || wanted[id] {
			return fail()
		}
		wanted[id] = true
	}
	type stock struct {
		id                             string
		holder                         PawnID
		amount, expiry, eligibleDemand float64
		eaters                         map[PawnID]bool
	}
	stocks := make([]stock, 0, len(supply.Stocks))
	seen := map[string]bool{}
	for _, input := range supply.Stocks {
		amount, ak := input.Nutrition.Value()
		holder, hk := input.Holder.Value()
		perishable, pk := input.Perishable.Value()
		if !foodID(input.ID) || seen[input.ID] || !ak || !foodNumber(amount) || !hk || !pk || len(input.Eaters) == 0 && !input.Corpse && !input.IsHumanMeat || len(input.Eaters) > len(ids) {
			return fail()
		}
		seen[input.ID] = true
		if holder != "" {
			if _, exists := demand[holder]; !exists || len(input.Eaters) != 1 || input.Eaters[0] != holder {
				return fail()
			}
		}
		entry := stock{id: input.ID, holder: holder, amount: amount, expiry: math.Inf(1), eaters: map[PawnID]bool{}}
		if perishable {
			ticks, known := input.RotTicks.Value()
			if !known || ticks < 0 {
				return fail()
			}
			entry.expiry = float64(ticks) / 60000
		}
		for _, id := range input.Eaters {
			if _, exists := demand[id]; !exists || entry.eaters[id] {
				return fail()
			}
			entry.eaters[id] = true
		}
		for _, id := range ids {
			if (input.IsHumanMeat || input.IsHumanlike) && !humanMeat[id] {
				delete(entry.eaters, id)
			}
			if entry.eaters[id] {
				entry.eligibleDemand += demand[id]
			}
		}
		if !foodNumber(entry.eligibleDemand) {
			return fail()
		}
		if input.Corpse {
			forbidden, known := input.Forbidden.Value()
			if !known {
				return fail()
			}
			if forbidden {
				continue
			}
		}
		if !input.Reserve {
			stocks = append(stocks, entry)
		}
	}
	sort.Slice(stocks, func(i, j int) bool {
		if stocks[i].expiry != stocks[j].expiry {
			return stocks[i].expiry < stocks[j].expiry
		}
		return stocks[i].id < stocks[j].id
	})
	result := FoodForecast{}
	for _, entry := range stocks {
		if entry.holder != "" && wanted[entry.holder] && entry.eaters[entry.holder] {
			result.InventoryNutrition += entry.amount
		}
	}
	for _, id := range ids {
		rate := demand[id]
		if rate == 0 || !wanted[id] {
			continue
		}
		row := ConsumerFoodForecast{ID: id, NutritionPerDay: rate}
		for _, entry := range stocks {
			if !entry.eaters[id] {
				continue
			}
			share := entry.amount * (rate / entry.eligibleDemand)
			if entry.holder == id {
				share = entry.amount
			}
			usable := math.Min(share, math.Max(0, entry.expiry-row.RunwayDays)*rate)
			row.AllocatedNutrition += share
			row.UsableNutrition += usable
			row.RunwayDays += usable / rate
		}
		if !foodNumber(row.RunwayDays) || !foodNumber(row.UsableNutrition) || !foodNumber(row.AllocatedNutrition) {
			return fail()
		}
		minimum, known := result.RunwayDays.Value()
		if !known || row.RunwayDays < minimum {
			result.RunwayDays = domain.Known(row.RunwayDays)
		}
		result.UsableNutrition += row.UsableNutrition
		result.AtRiskNutrition += math.Max(0, row.AllocatedNutrition-row.UsableNutrition)
		result.Consumers = append(result.Consumers, row)
	}
	if !foodNumber(result.UsableNutrition) || !foodNumber(result.AtRiskNutrition) || !foodNumber(result.InventoryNutrition) {
		return fail()
	}
	return result, nil
}
